src/memory_segment.rs

```diff
@@ -0,0 +1,256 @@
+//! Provide a [`BusDevice`] abstraction over a piece of a mmap'able
+//! file.
+
+use std::{
+    fs::File,
+    sync::{
+        atomic::{AtomicU16, AtomicU32, AtomicU64, AtomicU8, Ordering},
+        Arc,
+    },
+};
+
+use memmap2::{Mmap, MmapMut, MmapOptions};
+use usbvfiod::device::bus::{BusDevice, Request, RequestSize};
+use vfio_user::DmaMapFlags;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AccessRights {
+    ReadOnly,
+    ReadWrite,
+}
+
+impl TryFrom<DmaMapFlags> for AccessRights {
+    type Error = ();
+
+    fn try_from(value: DmaMapFlags) -> Result<Self, Self::Error> {
+        match value {
+            DmaMapFlags::READ_ONLY => Ok(AccessRights::ReadOnly),
+            DmaMapFlags::READ_WRITE => Ok(AccessRights::ReadWrite),
+            unsupported => todo!("unsupported DmaMapFlags {unsupported:?}"),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Mapping {
+    ReadOnly(Mmap),
+    ReadWrite(MmapMut),
+}
+
+impl Mapping {
+    fn as_ptr(&self) -> *const u8 {
+        match self {
+            Mapping::ReadOnly(map) => map.as_ptr(),
+            Mapping::ReadWrite(map) => map.as_ptr(),
+        }
+    }
+
+    fn is_writable(&self) -> bool {
+        match self {
+            Mapping::ReadWrite(_) => true,
+            Mapping::ReadOnly(_) => false,
+        }
+    }
+}
+
+/// A contiguous piece of mmap'ed memory.
+#[derive(Debug, Clone)]
+pub struct MemorySegment {
+    size: u64,
+    mapping: Arc<Mapping>,
+}
+
+impl MemorySegment {
+    /// Creates a memory segment from a file.
+    ///
+    /// The file is only used for mapping and will not be read or
+    /// written to.
```

Comment:
I found this a bit confusing with the `File` abstraction vs a file descriptor. We still read/write the (anonymous/memfd) "file" through the mapping.

```suggestion
    /// The File is only used for mapping and will not be read or
    /// written to.
```

Reply:
Good point. I reworded this.